src/main/java/org/gaul/s3proxy/crypto/DecryptionInputStream.java

```diff
@@ -0,0 +1,385 @@
+/*
+ * Copyright 2014-2021 Andrew Gaul <andrew@gaul.org>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gaul.s3proxy.crypto;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.TreeMap;
+
+import javax.annotation.concurrent.ThreadSafe;
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+import javax.crypto.ShortBufferException;
+
+import org.apache.commons.io.IOUtils;
+
+@ThreadSafe
+public class DecryptionInputStream extends FilterInputStream {
+
+    // the cipher engine to use to process stream data
+    private final Cipher cipher;
+
+    // the secret key
+    private final SecretKey key;
+
+    // the list of parts we expect in the stream
+    private final TreeMap<Integer, PartPadding> parts;
+
+    // the underlying input stream
+    private final InputStream input;
```

Comment:
Do you need to store this?  Can you just use the `FilterInputStream` superclass `in` field?

Reply:
you are totally right